### PR TITLE
Security HomWork

### PR DESCRIPTION
**Pull request by yuriikozar**
Solution + checkstyle
v1

### CODE REVIEW

File: src/main/java/taxi/controller/LoginController.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String username = req.getParameter("login");
+        String password = req.getParameter("password");
+        try {
+            Driver driver = authenticationService.login(username, password);
+            HttpSession session = req.getSession();
+            session.setAttribute("driver_id", driver.getId());
+            resp.sendRedirect("/index");
```
Comment (dSkarzh) [inline src/main/java/taxi/controller/LoginController.java:30]:
```suggestion
            resp.sendRedirect(req.getContextPath() + "/index");
```
please add req.getContextPath() everytime you use sendRedirect()

File: src/main/java/taxi/controller/driver/AddDriverController.java [hunk 2]
```diff
@@ -24,8 +24,12 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp)
     public void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
         String name = req.getParameter("name");
         String licenseNumber = req.getParameter("license_number");
+        String username = req.getParameter("login");
+        String password = req.getParameter("password");
         Driver driver = new Driver(name, licenseNumber);
+        driver.setPassword(password);
+        driver.setUsername(username);
```
Comment (dSkarzh) [inline src/main/java/taxi/controller/driver/AddDriverController.java:31]: let's use only constructors or setters, but not both

File: src/main/java/taxi/model/Driver.java [hunk 3]
```diff
@@ -3,6 +3,8 @@
 import java.util.Objects;
 
 public class Driver {
+    private String username;
+    private String password;
```
Comment (dSkarzh) [inline src/main/java/taxi/model/Driver.java:7]: add these fields to equals and hashcode

File: src/main/java/taxi/service/AuthenticationService.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package taxi.service;
+
+import taxi.exception.AuthenticationException;
+import taxi.model.Driver;
+
+public interface AuthenticationService {
+    public Driver login(String login, String password) throws AuthenticationException;
```
Comment (dSkarzh) [inline src/main/java/taxi/service/AuthenticationService.java:7]:
```suggestion
    Driver login(String login, String password) throws AuthenticationException;
```

File: src/main/java/taxi/service/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public Driver login(String login, String password) throws AuthenticationException {
+        Optional<Driver> driver = driverDao.findByLogin(login);
+        if (driver.isEmpty()) {
+            throw new AuthenticationException("Login or password was incorrect");
+        }
+        if (driver.get().getPassword().equals(password)) {
+            return driver.get();
+        }
+        throw new AuthenticationException("Login or password was incorrect");
```
Comment (dSkarzh) [inline src/main/java/taxi/service/AuthenticationServiceImpl.java:24]:
```suggestion
        if (driver.isEmpty() || !driver.get().getPassword().equals(password)) {
            throw new AuthenticationException("Login or password was incorrect");
        }
        return driver.get();
```

File: src/main/java/taxi/web/filter/AuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        allowedUrls.add("/drivers/add");
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        HttpSession session = req.getSession();
+
```
Comment (dSkarzh) [inline src/main/java/taxi/web/filter/AuthenticationFilter.java:31]:
```suggestion
```

File: src/main/java/taxi/controller/LogoutController.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet(urlPatterns = "/logout")
+public class LogoutController extends HttpServlet {
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
```
Comment (dSkarzh) [inline src/main/java/taxi/controller/LogoutController.java:13]:
```suggestion
    public void doGet(HttpServletRequest req, HttpServletResponse resp)
```
better to make these overridden methods public

**Review by dSkarzh — APPROVED** [review]

File: src/main/java/taxi/service/DriverService.java [hunk 8]
```diff
@@ -3,4 +3,5 @@
 import taxi.model.Driver;
 
 public interface DriverService extends GenericService<Driver> {
+    Driver findByLogin(String login);
```
Comment (Sofasmile) [inline src/main/java/taxi/service/DriverService.java:6]: цей метод на рівні сервісу повинен повертати optional

**Review by Sofasmile — CHANGES_REQUESTED** [review]

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**